### PR TITLE
core: 4844 opcode and precompile

### PR DESCRIPTION
**Pull request by MariusVanDerWijden**
This PR implements the pointEvaluation precompile and the DATAHASH opcode both are needed for 4844
Spec: https://eips.ethereum.org/EIPS/eip-4844#point-evaluation-precompile

### CODE REVIEW

File: crypto/kzg4844/kzg4844.go [hunk 1]
```diff
@@ -61,6 +63,8 @@ func UseCKZG(use bool) error {
 	} else {
 		gokzgIniter.Do(gokzgInit)
 	}
+	// Compute the precompile return value using go-kzg
```
Comment (karalabe) [inline crypto/kzg4844/kzg4844.go:66]: Please don't default to one or the other, if there's a bug in one of the libs, we made them swappable for a reason.

File: crypto/kzg4844/kzg4844.go [hunk 2]
```diff
@@ -41,6 +41,8 @@ type Point [32]byte
 // Claim is a claimed evaluation value in a specific point.
 type Claim [32]byte
 
+var PrecompileReturnValue [64]byte
```
Comment (karalabe) [inline crypto/kzg4844/kzg4844.go:44]: This really has no place in the crypto lib.

File: crypto/kzg4844/kzg4844_gokzg.go [hunk 3]
```diff
@@ -96,3 +97,12 @@ func gokzgVerifyBlobProof(blob Blob, commitment Commitment, proof Proof) error {
...
 }
+
+// goComputePrecompileReturnValue computes the precompile return value which is
+// FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values.
+func goComputePrecompileReturnValue() [64]byte {
+	var precompileReturnValue [64]byte
+	binary.BigEndian.PutUint64(precompileReturnValue[24:], gokzg4844.ScalarsPerBlob)
+	copy(precompileReturnValue[32:], gokzg4844.BlsModulus[:])
+	return precompileReturnValue
+}
```
Comment (karalabe) [inline crypto/kzg4844/kzg4844_gokzg.go:108]: Please move this out, there's no notion of "precompile" in a crypto library

File: core/vm/contracts.go [hunk 4]
```diff
@@ -1048,3 +1070,66 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
 	// Encode the G2 point to 256 bytes
 	return g.EncodePoint(r), nil
 }
+
+// pointEvaluation implements the EIP-4844 point evaluation precompile.
+type pointEvaluation struct{}
```
Comment (karalabe) [inline core/vm/contracts.go:1075]: Can we perhaps give this a more meaningful name throughout the code? It doesn't say much.

e.g. kzgPointEvaluation or something

File: core/vm/contracts.go [hunk 5]
```diff
@@ -90,6 +91,21 @@ var PrecompiledContractsBerlin = map[common.Address]PrecompiledContract{
 	common.BytesToAddress([]byte{9}): &blake2F{},
 }
 
+// PrecompiledContractsCancun contains the default set of pre-compiled Ethereum
+// contracts used in the Berlin release.
```
Comment (holiman) [inline core/vm/contracts.go:95]:
```suggestion
// contracts used in the Cancun release.
```

File: core/vm/contracts.go [hunk 6]
```diff
@@ -90,6 +91,21 @@ var PrecompiledContractsBerlin = map[common.Address]PrecompiledContract{
...
+	common.BytesToAddress([]byte{1}):  &ecrecover{},
+	common.BytesToAddress([]byte{2}):  &sha256hash{},
+	common.BytesToAddress([]byte{3}):  &ripemd160hash{},
+	common.BytesToAddress([]byte{4}):  &dataCopy{},
+	common.BytesToAddress([]byte{5}):  &bigModExp{eip2565: true},
+	common.BytesToAddress([]byte{6}):  &bn256AddIstanbul{},
+	common.BytesToAddress([]byte{7}):  &bn256ScalarMulIstanbul{},
+	common.BytesToAddress([]byte{8}):  &bn256PairingIstanbul{},
+	common.BytesToAddress([]byte{9}):  &blake2F{},
+	common.BytesToAddress([]byte{20}): &pointEvaluation{},
```
Comment (holiman) [inline core/vm/contracts.go:106]: Whaaa.. -- it's at `20`?   :eyes:

File: core/vm/contracts.go [hunk 7]
```diff
@@ -1048,3 +1070,66 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+
+	// Proof: next 48 bytes
+	var proof kzg4844.Proof
+	copy(proof[:], input[144:pointBlobVerifyInputLength])
+
+	if err := kzg4844.VerifyProof(commitment, point, claim, proof); err != nil {
+		return nil, errors.Join(errBlobVerifyKZGProof, err)
+	}
+
+	return common.CopyBytes(kzg4844.PrecompileReturnValue[:]), nil
```
[the remark this answers is not in the document]
Comment (holiman) [reply core/vm/contracts.go:1126]: I mean... what if an rpc call happens to invoke the precompile too, for example?

[the remark this answers is not in the document]
Comment (holiman) [reply core/vm/contracts.go:1126]: Ahh.. wait.. the return value of this op is not input-dependent??

File: core/vm/contracts.go (hunk 6 above)
Comment (MariusVanDerWijden) [reply core/vm/contracts.go:106]: Yep, because of the BLS precompiles

File: core/vm/contracts_test.go [hunk 8]
```diff
@@ -311,6 +313,10 @@ func TestPrecompiledBLS12381G2MultiExp(t *testing.T) { testJson("blsG2MultiExp",
 func TestPrecompiledBLS12381Pairing(t *testing.T)    { testJson("blsPairing", "10", t) }
 func TestPrecompiledBLS12381MapG1(t *testing.T)      { testJson("blsMapG1", "11", t) }
 func TestPrecompiledBLS12381MapG2(t *testing.T)      { testJson("blsMapG2", "12", t) }
+func TestPrecompiledPointEvaluation(t *testing.T) {
+	kzg4844.UseCKZG(false) // We need to call this to compute the PrecompileReturnValue
```
Comment (karalabe) [inline core/vm/contracts_test.go:317]: Nononononon. No dirty hacks

File: core/vm/contracts.go [hunk 9]
```diff
@@ -1048,3 +1070,67 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+		return nil, errors.Join(errBlobVerifyKZGProof, err)
+	}
+
+	return common.Hex2Bytes(blobPrecompileReturnValue), nil
+}
+
+// KZGToVersionedHash implements kzg_to_versioned_hash from EIP-4844
+func KZGToVersionedHash(kzg kzg4844.Commitment) common.Hash {
+	h := sha256.Sum256(kzg[:])
+	h[0] = blobVerifyKZGVersion
```
Comment (karalabe) [inline core/vm/contracts.go:1133]: params.BlobTxHashVersion ?

File: core/vm/contracts.go [hunk 10]
```diff
@@ -1048,3 +1070,67 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+	// input kzg point: next 48 bytes
+	var commitment kzg4844.Commitment
+	copy(commitment[:], input[96:144])
+	if KZGToVersionedHash(commitment) != versionedHash {
+		return nil, errBlobVerifyMismatchedVersion
+	}
+
+	// Proof: next 48 bytes
+	var proof kzg4844.Proof
+	copy(proof[:], input[144:blobVerifyInputLength])
```
Comment (karalabe) [inline core/vm/contracts.go:1121]: I think throughout you don't need to specify an end-range for the slicings. Copy will only use as much data as space in the destination and all the kzg4844 types are of fixed sizes.

File: core/vm/contracts.go [hunk 11]
```diff
@@ -1048,3 +1070,67 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+
+	if err := kzg4844.VerifyProof(commitment, point, claim, proof); err != nil {
+		return nil, errors.Join(errBlobVerifyKZGProof, err)
```
Comment (karalabe) [inline core/vm/contracts.go:1124]: Ugh, this is such a complex method. It's meant to join many errors. Here let's just use something simple:

`fmt.Errorf("%w: %v", errBlobVerifyKZGProof, err))`

File: core/vm/contracts.go [hunk 12]
```diff
@@ -1048,3 +1070,67 @@ func (c *bls12381MapG2) Run(input []byte) ([]byte, error) {
...
+	}
+
+	return common.Hex2Bytes(blobPrecompileReturnValue), nil
+}
+
+// KZGToVersionedHash implements kzg_to_versioned_hash from EIP-4844
+func KZGToVersionedHash(kzg kzg4844.Commitment) common.Hash {
```
Comment (karalabe) [inline core/vm/contracts.go:1131]: Make this private?

File: core/vm/contracts.go (hunk 9 above)
Comment (lightclient) [reply core/vm/contracts.go:1133]: `params.BlobCommitmentVersionKZG`?

(this is how it is in the EIP)

File: core/vm/contracts.go [hunk 13]
```diff
@@ -1081,7 +1082,7 @@ func (b *kzgPointEvaluation) RequiredGas(input []byte) uint64 {
 
 const (
 	blobVerifyInputLength           = 192  // Max input length for the point evaluation precompile.
-	blobVerifyKZGVersion      uint8 = 0x01 // Version byte for the point evaluation precompile.
+	blobCommitmentVersionKZG  uint8 = 0x01 // Version byte for the point evaluation precompile.
```
Comment (karalabe) [inline core/vm/contracts.go:1085]: This is part of the params package, shouldn't we use that?

**Review by karalabe — APPROVED** [review]
SGTM

**Pull request merged**